forest/src/daemon.rs

```diff
@@ -151,10 +155,22 @@ pub(super) async fn start(config: Config) {
     .unwrap();
     chain_store.set_genesis(&genesis.blocks()[0]).unwrap();
 
+    // Reward calculation is needed by the VM to calculate state, which can happen essentially anywhere the `StateManager` is called.
+    // It is consensus specific, but threading it through the type system would be a nightmare, which is why dynamic dispatch is used.
+    #[cfg(all(feature = "fil_cns", not(any(feature = "deleg_cns"))))]
+    let reward_calc = fil_cns::reward_calc();
+    #[cfg(feature = "deleg_cns")]
+    let reward_calc = deleg_cns::reward_calc();
```

Comment:
I agree with @LesnyRumcajs to have the consensus built as a runtime feature.

Reply:
Thanks for advising on this issue. I'm not happy about all the noise of setting up the different consensus types in `daemon.rs` and thought about moving the stuff into their respective modules to only keep a single switch here. It would be more boilerplate in the consensus modules because they could no longer capture variables from the code, but being explicit would most likely be a good thing in this case.

I agree that dynamic dispatch shouldn't be a problem in this case because rewards are calculated once per block. Validation and block creation are similarly coarse grained. I thought maybe the tipset weight calculation benefits from being statically dispatched as it could be called multiple times, and it doesn't have an instance either, so there's nothing to dispatch on at the moment.

But I'm not sure if you would want to compile all kinds of experimental consensus types which are more heavy than this Delegated Consensus into every Forest binary. They can vary wildly in their dependencies; in the current [examples](https://github.com/filecoin-project/eudico/tree/eudico/chain/consensus) there are ones that talk to Tendermint, others do their own networking. The vision is that a parent subnet should not necessarily be aware of what consensus the child subnet is running, and wanted to avoid sleepwalking into [harcoding the possible types](https://github.com/filecoin-project/eudico/blob/f4c54d4ef2c0b157fbc535cd9e4afabe2d9cdf54/chain/consensus/hierarchical/types.go#L18) in the Forest codebase.

Let me know what you think!

Reply:
I moved consensus initialisation out of `daemon.rs`, so you see what it would look like.

Reply:
@aakoshh Thanks for the explanation, it makes sense now.